src/foam/nanos/mrac/ClusterDAO.js

```diff
@@ -0,0 +1,182 @@
+  2. creates Client DAOs.
+  3. on put() write through to delegate (MDAO), then send to Clients.
+  4. recreate clients on configuration changes.
+  `,
+
+  javaImports: [
+    'foam.dao.DAO',
+    'foam.dao.ArraySink',
+    'static foam.mlang.MLang.*',
+    'foam.nanos.boot.NSpec',
+    'foam.nanos.logger.Logger',
+    'java.util.List',
+    'java.util.ArrayList',
+  ],
+
+  implements: [
+    'foam.nanos.boot.NSpecAware'
+  ],
+
+  properties: [
+    {
+      // REVIEW - this may no longer be required
+      documentation: `nSpec of the DAO to be clustered.`,
+      name: 'nSpec',
+      class: 'FObjectProperty',
+      type: 'foam.nanos.boot.NSpec',
+      visibility: 'HIDDEN'
+    },
+    {
+      documentation: `Cluster configuration for 'this' (localhost) node.`,
+      name: 'config',
+      class: 'FObjectProperty',
+      type: 'foam.nanos.mrac.ClusterConfig',
+      visibility: 'RO'
+    },
+    {
+      documentation: `Array of all clients to put/remove to after a non-server operation.`,
+      name: 'clients',
+      class: 'FObjectArray',
+      of: 'foam.dao.DAO',
+      visibility: 'HIDDEN'
+    },
+    {
+      documentation: `Reference to the ClusterConfigDAO to which we are subscribed/listening for updates to reconfigure.`,
+      name: 'clusterConfigDAO',
+      class: 'foam.dao.DAOProperty',
+      visibility: 'HIDDEN'
+    }
+  ],
+
+  methods: [
+    {
+      documentation: `Upon initialization create the ClusterServer configuration and register nSpec.`,
+      name: 'init_',
+      javaCode: `
+        reconfigure(getX());
+
+        /* register ClusterConfig Listener */
+        DAO clusterConfigDAO = (DAO) getX().get("clusterConfigDAO");
+        clusterConfigDAO.listen(new ClusterConfigSink(getX(), this), TRUE);
+        /* REVIEW: need to keep a reference to the dao?*/
+        setClusterConfigDAO(clusterConfigDAO);
+      `
+    },
+    {
+      documentation: `Rebuild the client list.`,
+      name: 'reconfigure',
+      args: [
+        {
+          name: 'x',
+          type: 'Context'
+        }
+      ],
+      javaCode: `
+      Logger logger = (Logger) x.get("logger");
+      logger.debug(this.getClass().getSimpleName(), "reconfigure", getNSpec().getName());
+      DAO dao = (DAO) x.get("clusterConfigDAO");
+      List arr = (ArrayList) ((ArraySink) dao
+       .where(
+         AND(
+           EQ(ClusterConfig.ENABLED, true),
+           EQ(ClusterConfig.STATUS, Status.ONLINE)
+         )
+       )
+       .select(new ArraySink())).getArray();
+      DAO[] newClients = new DAO[arr.size()];
+      for ( int i = 0; i < arr.size(); i++ ) {
+        ClusterConfig config = (ClusterConfig) arr.get(i);
+        if ( "localhost".equals(config.getId()) ) {
+          setConfig(config);
+          continue;
+        }
+        DAO client = new ClusterClientDAO.Builder(x).setServiceName(getNSpec().getName()).setConfig(config).build();
```

Comment:
Instead of creating a ClusterClientDAO here we can create a ClientDAO and call setDelegate(new HTTPBox()) like we did in our HelloService test case.